src/transformers/models/superglue/modeling_superglue.py

```diff
@@ -0,0 +1,804 @@
+        hidden_states (`tuple(torch.FloatTensor)`, *optional*, returned when `output_hidden_states=True` is passed or
+        when `config.output_hidden_states=True`):
+            Tuple of `torch.FloatTensor` (one for the output of each stage) of shape `(batch_size, 2, num_channels,
+            num_keypoints)`.
+        attentions (`tuple(torch.FloatTensor)`, *optional*, returned when `output_attentions=True` is passed or when
+        `config.output_attentions=True`):
+            Tuple of `torch.FloatTensor` (one for each layer) of shape `(batch_size, 2, num_heads, num_keypoints,
+            num_keypoints)`.
+    """
+
+    loss: Optional[torch.FloatTensor] = None
+    mask: torch.FloatTensor = None
+    matches: torch.FloatTensor = None
+    matching_scores: torch.FloatTensor = None
+    keypoints: torch.FloatTensor = None
+    hidden_states: Optional[Tuple[torch.FloatTensor]] = None
+    attentions: Optional[Tuple[torch.FloatTensor]] = None
+
+
+class SuperGlueMultiLayerPerceptron(nn.Module):
+    def __init__(self, config: SuperGlueConfig, channels: List[int]) -> None:
+        super().__init__()
+        num_layers = len(channels)
+        layers = []
+        for i in range(1, num_layers):
+            layers.append(nn.Conv1d(channels[i - 1], channels[i], kernel_size=1))
+            if i < (num_layers - 1):
+                layers.append(nn.BatchNorm1d(channels[i]))
+                layers.append(nn.ReLU())
+        self.layers = nn.Sequential(*layers)
+
+    def forward(
+        self, hidden_state: torch.Tensor, output_hidden_states: Optional[bool] = False
+    ) -> Tuple[torch.Tensor, Optional[Tuple[torch.Tensor]]]:
+        all_hidden_states = () if output_hidden_states else None
+        for layer in self.layers:
+            hidden_state = layer(hidden_state)
+            if output_hidden_states and isinstance(layer, nn.Conv1d):
+                all_hidden_states = all_hidden_states + (hidden_state,)
+        return hidden_state, all_hidden_states
+
+
+class SuperGlueKeypointEncoder(nn.Module):
+    def __init__(self, config: SuperGlueConfig) -> None:
+        super().__init__()
+        layer_sizes = config.keypoint_encoder_sizes
+        feature_dim = config.descriptor_dim
+        # 3 here consists of 2 for the (x, y) coordinates and 1 for the score of the keypoint
+        encoder_channels = [3] + layer_sizes + [feature_dim]
+        self.encoder = SuperGlueMultiLayerPerceptron(config, channels=encoder_channels)
+
+    def forward(
+        self,
+        keypoints: torch.Tensor,
+        scores: torch.Tensor,
+        output_hidden_states: Optional[bool] = False,
+    ) -> Tuple[torch.Tensor, Optional[Tuple[torch.Tensor]]]:
+        keypoints = keypoints.transpose(1, 2)
+        scores = scores.unsqueeze(1)
+        inputs = torch.cat([keypoints, scores], dim=1)
+        return self.encoder(inputs, output_hidden_states=output_hidden_states)
+
+
+class SuperGlueMultiHeadAttention(nn.Module):
```

Comment:
I managed to convert all SuperGlue's use of Conv1d layers instead of Linear layers. The Attention module now has a more "standard implementation". Although, I couldn't find another Attention class which only implement what SuperGlue uses, I didn't want to copy classes that made use of masking or layer pruning for example as it is not required in SuperGlue. Do you have a class in mind which I should copy ?

Reply:
Should I implement it with the modular guidelines instead of copied from ?